examples/r-future/r-furrr.Rmd

```diff
@@ -0,0 +1,210 @@
+        data = as.matrix(test_data),
+        label = test_df$weight_pounds
+    )
+
+    return(list("train" = dtrain, "test" = dtest))
+}
+```
+
+### Train and Test the Model
+Our last function definitions define and train the model and determine the quality of the results. The model is fit, the mean absolute error calculated, and the results added to a table with the corresponding hyperparameters.
+
+> **Note**: We disable XGBoost's multi-threading in this example to show the difference between furrr and purrr for single-threaded workloads.
+
+```{r train model}
+train_model <- function(params, dtrain) {
+    model <- xgb.train(
+        params = params,
+        data = dtrain,
+        nrounds = 100,
+        nthread = 1,
+        objective = "reg:squarederror",
+    )
+    return(model)
+}
+
+test_results <- function(model, dtest) {
+    results <- predict(model, dtest)
+    return(results)
+}
+
+create_model_table <- function(data, ...) {
+    dmatrices <- create_matrices(data)
+
+    dtrain <- dmatrices$train
+    dtest <- dmatrices$test
+
+    params <- list(...)
+
+    model <- train_model(params, dtrain)
+    results <- test_results(model, dtest)
+    return(
+        tibble(
+            mean_absolute_error = mae(getinfo(dtest, "label"), results),
+            params = params
+        )
+    )
+}
+```
+
+### Let's Run the Model!
+It's finally time to run the functions we just created. We also define the hyperparameters we want to test. We use the expand_grid function (with `list()` around constants) to create the appropriate. Here we are only testing max_depth, but other hyperparameters could be added to the grid.
+
+Here are all the preprocessing steps.
+
+```{r preprocess}
+births_raw_data <- download_data()
+births_data <- filter_data(births_raw_data)
+births_data_preprocessed <- preprocess_data(births_data)
+
+params <- expand_grid(
+    data = list(births_data_preprocessed),
+    max_depth = seq(1, 8)
+)
+```
+
+#### Try Running the Model with purrr
+Let's first try it with purrr to iterate through the various models.
+
+Actually, let's not -- it takes over 17 minutes to run this code.  
+```{r purrr evaluation, eval = FALSE}
+
+# ###################################################
+# ## Don't actually run this - it takes forever... ##
+# ###################################################
+
+# print("With purrr:")
+# tic()
+# results_purrr <- pmap_dfr(params, create_model_table)
+# toc()
+
+# best_result_purrr <- results_purrr %>%
+#     top_n(-1, mean_absolute_error) %>%
+#     head(1)
+
+# print(best_result_purrr)
+```
+
+#### Run the Model with furrr
+
+Instead, let's try it with a parallel back-end. In this case, we ask for eight workers for eight processes.
+
+The only change to the purrr code is to add the futures `plan` and exchanging `pmap_dfr` with `future_pmap_dfr`. 
+
+```{r furrr evaluation}
+#####################
+## Run this instead##
+#####################
+plan(multisession(workers = 8))
+
+print("With furrr:")
```

Comment:
```suggestion
message("With furrr:")
```

In general it's better to use `message` than `print` for like, logging progress. Message is better because you can use commands like `suppressMessages()` and it's handled differently sometimes when other functions are calling subfunctions.

(there are a few other places you use print too that you might want to swtich)